### PR TITLE
Add regular expression support for content checking

### LINKED ISSUE #102: Regular Expressions for Capturing Fuller Fragments
@gkapfham brought up the idea of creating code with more in-depth fragment capture (for expressions). Specifically, he mentioned that a single fragment may not be able to analyze the entirety of certain LaTeX elements such as an enumerated list. Finally, adding this check and those that are similar would allow the system to be useful in additional situations.

### PR DESCRIPTION
**Pull request by burrowss**
We have implemented a regex checker for GatorGrader using a command line argument --regex

The features include new functions to check whether the input from the user is a valid regex or not, and if it is, it then checks the files for the specified regex. We have also added test cases to check the functions added. Documentation has been added to accurately reflect how this feature can be used.

Fixes #102.

### CODE REVIEW

**Comment by Michionlion** [comment]
Hi @burrowss, more details need to be added to your PR. Things like, how does one use this check, what does it change in regards to other checks, etc. should be included.

File: Pipfile [hunk 1]
```diff
@@ -6,6 +6,9 @@ name = "pypi"
 [packages]
 gitpython = "*"
 requests = ">=2.20.0"
+pytest = "*"
+"flake8" = "*"
```
Comment (Michionlion) [inline Pipfile]: These are added to an incorrect section of the Pipfile. This ***must*** be fixed -- these are not run-time dependencies, but development dependencies!

File: README.md [hunk 2]
```diff
@@ -226,7 +227,7 @@ the following error message `gatorgrader.py: error: unrecognized arguments:
 pytest --count 1 --exact` uses the correct names for all of the specified
 arguments, it does not tell GatorGrader what type of deliverable to check (e.g.,
 a program's source code or a command's output), thus leading GatorGrader to
-report `Incorrect command-line arguments`.
+report `Incorrect command-line arguments`. Similar errors will occur for the incorrect use of the `--regex` command. One correct usage of the `--regex` command would be `--regex "\\begin([^;]*)\\end" --count 1` to count amount of the `\begin` and `\end` regular expressions in a LaTeX document. Similarly, `--regex "\<footer\>([^;]*)\</footer\>" --count 1` could be used to count amount of regex for footers in an HTML file.
```
Comment (Michionlion) [inline README.md]: This line must be cut so that the lines match the correct line length. I'm not sure why `mdl` does not catch this.

File: gator/invoke.py [hunk 3]
```diff
@@ -285,6 +285,80 @@ def invoke_all_fragment_checks(
...
+        if exact is not True:
+            message = (
+                "The "
+                + filecheck
+                + " in "
+                + directory
+                + " has at least "
+                + str(expected_count)
+                + " of the '"
+                + fragment
```
Comment (Michionlion) [inline gator/invoke.py]: The variable should be called regex and this message should probably mention "matching", you can't simply reuse `fragment`'s message.

File: gator/invoke.py [hunk 4]
```diff
@@ -285,6 +285,80 @@ def invoke_all_fragment_checks(
...
+                + " has exactly "
+                + str(expected_count)
+                + " of the '"
+                + fragment
+                + "' regular expression"
+            )
+    diagnostic = (
+        "Found "
+        + str(actual_count)
+        + " regular expression(s) in the output or the specified file"
```
Comment (Michionlion) [inline gator/invoke.py]: Again, this should mention "matchings", not "regular expressions(s)", that doesn't make sense.

File: gator/invoke.py [hunk 5]
```diff
@@ -285,6 +285,80 @@ def invoke_all_fragment_checks(
...
+    fragment,
+    expected_count,
+    filecheck=NOTHING,
+    directory=NOTHING,
+    contents=NOTHING,
+    exact=False,
+):
+    """Perform the check for a regex existence in file or contents and return the results"""
+    met_or_exceeded_count = 0
+    met_or_exceeded_count, actual_count = fragments.specified_fragment_greater_than_count(
```
Comment (Michionlion) [inline gator/invoke.py]: If you're reusing the fragment method, it should probably be refactored (since fragment is a specialized version of the regex check). It's name should be something like `specified_entity_greater_than_count`.

File: gator/orchestrate.py [hunk 6]
```diff
@@ -184,6 +184,27 @@ def check_fragment_file(system_arguments):
     return actions
 
 
+def check_regex_file(system_arguments):
+    """Check the existence of fragment in a file and return desired actions"""
```
Comment (Michionlion) [inline gator/orchestrate.py]: This docstring must be updated.

File: tests/test_fragments.py [hunk 7]
```diff
@@ -378,3 +378,20 @@ def test_detect_blank_line(writing_string, expected_status):
     """Create some strings and then see blank detection for the line works"""
     found_blanks = fragments.is_blank_line(writing_string)
     assert found_blanks == expected_status
+
+
+@pytest.mark.parametrize(
+    "regex_fragment,expected_status", [("test", True), ("[", False)]
```
Comment (Michionlion) [inline tests/test_fragments.py]: More test cases should be added here to verify the correct working.

File: tests/test_fragments.py [hunk 8]
```diff
@@ -378,3 +378,20 @@ def test_detect_blank_line(writing_string, expected_status):
...
+)
+def test_is_valid_regex(regex_fragment, expected_status):
+    """Checks that regex assessment correctly validates regular expressions"""
+    regex_status = fragments.is_valid_regex(regex_fragment)
+    assert regex_status == expected_status
+
+
+@pytest.mark.parametrize("regex_fragment,expected_status", [("[", True)])
```
Comment (Michionlion) [inline tests/test_fragments.py]: Many more test cases are needed here!
This is not even verifying that the count is correct!

You should look to the fragment tests as examples.

File: tests/test_invoke.py [hunk 9]
```diff
@@ -196,13 +199,18 @@ def test_content_string_check_fragments(reset_results_dictionary):
 # pylint: disable=unused-argument
 # pylint: disable=redefined-outer-name
 def test_content_string_check_fragments_exact(reset_results_dictionary):
-    """Checks that the checking of words works correctly"""
+    """Checks that the checking of exact fragments and regex works correctly"""
```
Comment (Michionlion) [inline tests/test_invoke.py]: Your tests should ***never*** be added to an unrelated test case. This MUST be a new method!!

**Comment by burrowss** [comment]
I will have my team review these changes before we continue with the pull request. Thank you for the feedback

**Comment by Michionlion** [comment]
You should continue to use this PR for review and updates - all the relevant fields can be updated, and new commits to your branch will show here.

**Comment by lussierc** [comment]
Here are some sample uses of the regex support:
- Counting footer regex in HTML: `--regex "\<footer\>([^;]*)\</footer\>" --count 1`
- Counting `\begin` and `\end` regex in LaTeX: `--regex "\\begin([^;]*)\\end" --count 1`

File: Pipfile (hunk 1 above)
Comment (lussierc) [reply Pipfile]: @Michionlion I have removed these lines of code, I don't really think they were necessary as they already appear in `[dev-packages]`.

File: gator/fragments.py [hunk 10]
```diff
@@ -128,13 +138,13 @@ def specified_fragment_greater_than_count(
     contents=NOTHING,
     exact=False,
 ):
-    """Determines if the fragment count is greater than expected"""
-    # count the fragments in either a file in a directory or String contents
-    file_fragment_count = count_fragments(
+    """Determines if the entity count is greater than expected"""
+    # count the fragments/regex in either a file in a directory or String contents
+    file_entity_count = count_fragments(
```
Comment (Michionlion) [inline gator/fragments.py]: `count_fragments` should also be fixed to express its use here -- `count_entities` may be better.

File: gator/invoke.py (hunk 3 above)
Comment (Michionlion) [reply gator/invoke.py]: @everitt-andrew the variable should also be renamed so that future developers are not confused.

File: gator/fragments.py (hunk 10 above)
Comment (lussierc) [reply gator/fragments.py]: @Michionlion Thanks for the suggestion, I have made this change across the necessary files.

File: tests/test_fragments.py (hunk 8 above)
Comment (everitt-andrew) [reply tests/test_fragments.py]: @leonardoz15 is working to finish these soon and will resolve the conversation once he has done so.

**Comment by leonardoz15** [comment]
I wrote more test cases to test the regex counting in `test_fragments.py`, imitating what was written for the fragment counting. Is there anything else I should change about these tests ?

File: tests/test_fragments.py [hunk 11]
```diff
@@ -378,3 +378,127 @@ def test_detect_blank_line(writing_string, expected_status):
...
+def test_is_valid_regex(regex_fragment, expected_status):
+    """Checks that regex assessment correctly validates regular expressions"""
+    regex_status = fragments.is_valid_regex(regex_fragment)
+    assert regex_status == expected_status
+
+
+@pytest.mark.parametrize(
+    "writing_string,chosen_regex,expected_count",
+    [
+        ("hello world!!%^(@after)writing a lot\n\nnew one", "writing", 1),
```
Comment (Michionlion) [inline tests/test_fragments.py]: These tests must actually use a regex... otherwise what is this testing??

File: tests/test_fragments.py [hunk 12]
```diff
@@ -378,3 +378,127 @@ def test_detect_blank_line(writing_string, expected_status):
...
+        ("hello @world!!%^(@after)writing a lot\n\nnew one", "@world", 1),
+        ("hello world!!%^(@after)writing a lot\n\nnew one", "@world", 0),
+        ("[", ")", 0),
```
Comment (Michionlion) [inline tests/test_fragments.py]: This is an invalid regex.

**Comment by Michionlion** [comment]
@burrowss I cannot merge this PR until the problems I mentioned in my earlier review are fixed. Could you attempt to finish these by Wednesday morning? @gkapfham and I will be reviewing PRs for a final merge then.

**Comment by burrowss** [comment]
> @burrowss I cannot merge this PR until the problems I mentioned in my earlier review are fixed. Could you attempt to finish these by Wednesday morning? @gkapfham and I will be reviewing PRs for a final merge then.

@Michionlion  Sure, we will address these issues.  Thanks for the heads up

**Comment by leonardoz15** [comment]
I'm having trouble coming up with examples to test for multiple regular expressions in a string. Any suggestions?

**Comment by Michionlion** [comment]
@leonardoz15 @burrowss You should make sure all of the python strings that represent regexes are "raw" -- precededed by an `r`. This will make sure that escapes are respected correctly.

File: gator/fragments.py [hunk 13]
```diff
@@ -123,8 +123,22 @@ def count_specified_regex(contents, regex):
...
+    position_end = len(new_contents)
+
+    while is_match:
+        match = re.search(regex, repr(new_contents[position_start:position_end]))
+        if match is not None:
+            num_matches += 1
+            position_start = match.start() + 1
+            position_end = match.end() - 1
+        else:
+            is_match = False
```
Comment (Michionlion) [inline gator/fragments.py]: Why was this changed? It seems the last version was less complex and still should achieve the result.

Comment (burrowss) [reply gator/fragments.py]: We added this in so that it can iterate through multiple regex. As it was, if there were multiple regex, such as multiple \begins and \ends, it could not get the interior ones thus resulting in not being able to capture more than one of the regex

Comment (Michionlion) [reply gator/fragments.py]: Um... unless I'm misunderstanding, that's a failure of the regex, not of the algorithm. I'm not sure I understand though.

Comment (burrowss) [reply gator/fragments.py]: For example, in LaTeX, a potential regex could be:
\begin{enumerate}
    hello there
\end{enumerate}
\begin{enumerate}
    hello world
\end{enumerate}

Previously, if you want to check the amount of begin and ends, it would only find the first \begin and the last \end, resulting in only 1 regex for \begin{enumerate} and \end{enumerate} found. Now with these changes, it finds each instance, so it finds both \begin{enumerate} and \end{enumerates}, resulting in the correct 2 instances for the regex.

I hope this explanation helps. If not, let me know.

Comment (Michionlion) [reply gator/fragments.py]: Yes, that is a problem with the regex. You need to use lazy qualifiers ( `?` ) to ensure that the **first** instance of the characters following a `.*` is picked. You should not modify the algorithm for that -- regexes will now not be matched correctly.

Comment (leonardoz15) [reply gator/fragments.py]: I see what you mean, if we specify the regex as `\\begin([^;]*?)\\end` it will distinguish multiple groups of regex, like in Simon's example. 
The other problem is with regex examples embedded inside another regex. For example:
\begin{document}
hello
\begin{enumerate}
\item hello
\end{enumerate}
\end{document}

The algorithm doesn't see to distinguish between the first and second `\end`. Not sure if this is a fully unrelated problem but I feel like it applies to the kind of checks that would be performed on an actual LaTeX file.

Comment (Michionlion) [reply gator/fragments.py]: I'm not sure what you mean by `regex` here -- `\\begin([^;]*?)\\end` is a regex, the text is not.
The regex shouldn't distinguish between the two `\end`s -- there is nothing in the regex to do that! A positive lookahead with a backreference to a capturing group for the type of environment would work here. Perhaps brushing up on regular expressions using a resource like [regex 101](https://regex101.com/) would help.

As a side note, the regex you gave doesn't really make sense to detect begin-end blocks; the contents are simply everything that isn't `;`, however that could easily be part of the contents in LaTeX.

**Comment by Michionlion** [comment]
@gkapfham @burrowss I've made some changes and this should be ready to merge. @gkapfham can you just check through the changes to make sure I didn't miss anything?

**Comment by burrowss** [comment]
@Michionlion Thank you for the help! It looks like everything got covered
